src/os/bluestore/BitmapFreelistManager.cc

```diff
@@ -0,0 +1,413 @@
+    enumerate_bl_pos = get_next_clear_bit(enumerate_bl, enumerate_bl_pos);
+    if (enumerate_bl_pos >= 0) {
+      *offset = get_offset(enumerate_offset, enumerate_bl_pos);
+      dout(30) << __func__ << " found clear bit, key " << enumerate_offset
+	       << " bit " << enumerate_bl_pos
+	       << " offset " << *offset
+	       << dendl;
+      break;
+    }
+    dout(30) << " no more clear bits in " << enumerate_offset << dendl;
+    enumerate_p->next();
+    enumerate_bl.clear();
+    if (!enumerate_p->valid()) {
+      enumerate_offset += bytes_per_block * blocks_per_key;
+      enumerate_bl_pos = 0;
+      *offset = get_offset(enumerate_offset, enumerate_bl_pos);
+      break;
+    }
+    string k = enumerate_p->key();
+    const char *p = k.c_str();
+    uint64_t next = enumerate_offset + bytes_per_block * blocks_per_key;
+    _key_decode_u64(p, &enumerate_offset);
+    enumerate_bl = enumerate_p->value();
+    enumerate_bl_pos = 0;
+    if (enumerate_offset > next) {
+      dout(30) << " no key at " << next << ", got " << enumerate_offset << dendl;
+      *offset = next;
+      break;
+    }
+  }
+
+  // skip clear bits to find the end
+  uint64_t end = 0;
+  if (enumerate_p->valid()) {
+    while (true) {
+      enumerate_bl_pos = get_next_set_bit(enumerate_bl, enumerate_bl_pos);
+      if (enumerate_bl_pos >= 0) {
+	end = get_offset(enumerate_offset, enumerate_bl_pos);
+	dout(30) << __func__ << " found set bit, key " << enumerate_offset
+		 << " bit " << enumerate_bl_pos
+		 << " offset " << end
+		 << dendl;
+	*length = end - *offset;
+	dout(10) << __func__ << " " << *offset << "~" << *length << dendl;
+	return true;
+      }
+      dout(30) << " no more set bits in " << enumerate_offset << dendl;
+      enumerate_p->next();
+      enumerate_bl.clear();
+      enumerate_bl_pos = 0;
+      if (!enumerate_p->valid()) {
+	break;
+      }
+      string k = enumerate_p->key();
+      const char *p = k.c_str();
+      _key_decode_u64(p, &enumerate_offset);
+      enumerate_bl = enumerate_p->value();
+    }
+  }
+
+  end = size;
+  if (enumerate_offset < end) {
+    *length = end - enumerate_offset;
+    dout(10) << __func__ << " " << *offset << "~" << *length << dendl;
+    enumerate_offset = end;
+    return true;
+  }
+
+  dout(10) << __func__ << " end" << dendl;
+  return false;
+}
+
+void BitmapFreelistManager::dump()
+{
+  enumerate_reset();
+  uint64_t offset, length;
+  while (enumerate_next(&offset, &length)) {
+    dout(20) << __func__ << " " << offset << "~" << length << dendl;
+  }
+}
+
+int BitmapFreelistManager::allocate(
+  uint64_t offset, uint64_t length,
+  KeyValueDB::Transaction txn)
+{
+  dout(10) << __func__ << " " << offset << "~" << length << dendl;
+  return _xor(offset, length, txn);
+}
+
+int BitmapFreelistManager::release(
+  uint64_t offset, uint64_t length,
+  KeyValueDB::Transaction txn)
+{
+  dout(10) << __func__ << " " << offset << "~" << length << dendl;
+  return _xor(offset, length, txn);
+}
+
```

Comment:
Ditto here, under a debug option we should do a "get" and verify the state of the affected bits.